golem/managers/payment/pay_all.py

```diff
@@ -30,116 +34,119 @@ def __init__(
+    @trace_span("Stopping PayAllPaymentManager", log_level=logging.INFO)
+    async def stop(self):
+        """Terminate all related agreements."""
+        wait_task = asyncio.create_task(asyncio.sleep(self._shutdown_timeout))
+        wait_for_invoices_task = asyncio.create_task(self._wait_for_invoices())
+        _, pending = await asyncio.wait(
+            [wait_task, wait_for_invoices_task], return_when=asyncio.FIRST_COMPLETED
+        )
+        await ensure_cancelled_many(pending)
+
+        await asyncio.gather(*[agreement.close_all() for agreement in self._agreements.values()])
 
-    @trace_span("Getting allocation", show_results=True, log_level=logging.INFO)
-    async def get_allocation(self) -> Allocation:
-        # TODO handle NoMatchingAccount
-        if self._allocation is None:
-            await self._create_allocation()
+        for event_handler in self._event_handlers:
+            await self._golem.event_bus.off(event_handler)
 
-        # TODO fix type
-        return self._allocation  # type: ignore[return-value]
+        await self._release_allocation()
 
     @trace_span("Waiting for invoices", log_level=logging.INFO)
-    async def wait_for_invoices(self):
-        for _ in range(60):
-            if (
-                self._opened_agreements_count
-                == self._closed_agreements_count
-                == self._payed_invoices_count
-            ):
-                logger.info("All invoices paid")
-                return
-
-            await asyncio.sleep(1)
-
-        # TODO: Add list of agreements without payment
-        raise ManagerException("Waiting for invoices failed with timeout!")
-
-    async def _increment_opened_agreements(self, event: NewAgreement):
-        self._opened_agreements_count += 1
-
-    async def _increment_closed_agreements(self, event: AgreementClosed):
-        self._closed_agreements_count += 1
-
-    @trace_span(show_arguments=True)
-    async def _accept_invoice(self, invoice: Invoice) -> None:
-        assert self._allocation is not None  # TODO think of a better way
-        await invoice.accept_full(self._allocation)
-        await invoice.get_data(force=True)
-        self._payed_invoices_count += 1
-
-        logger.info(f"Invoice `{invoice.id}` for agreement `{invoice.data.agreement_id}` accepted")
-
-    @trace_span(show_arguments=True)
-    async def _accept_debit_note(self, debit_note: DebitNote) -> None:
-        assert self._allocation is not None  # TODO think of a better way
-        await debit_note.accept_full(self._allocation)
-        await debit_note.get_data(force=True)
-
-        logger.info(
-            f"DebitNote `{debit_note.id}` for agreement `{debit_note.data.agreement_id}` accepted"
-        )
+    async def _wait_for_invoices(self):
+        try:
+            while len(self._agreements) > 0:
+                await self._agreement_event.wait()
+                self._agreement_event.clear()
```

Comment:
The use of `self._agreement_event` looks suboptimal. As this event is only used for waiting for all invoices, instead of waking up every time an agreement is saved / discarded, it should be used to mark a desired state, e.g.: `self._no_agreements_event`. The state of that event should be handled in save (always clearing) / discard (setting conditionally if the last agreement was just removed).

Reply:
I don't understand

Reply:
Done